connectors/kafka/src/main/java/forklift/producers/KafkaForkliftProducer.java

```diff
@@ -0,0 +1,257 @@
+        return this.properties;
+    }
+
+    @Override
+    public void setProperties(Map<String, String> properties) throws ProducerException {
+        this.properties = properties;
+    }
+
+    @Override
+    public void setHeaders(Map<Header, Object> haeders) {
+        throw new UnsupportedOperationException("Kafka Producer does not support headers or properties");
+    }
+
+    @Override
+    public Map<Header, Object> getHeaders() {
+        throw new UnsupportedOperationException("Kafka Producer does not support headers or properties");
+    }
+
+    @Override
+    public void close() throws IOException {
+        //do nothing, the passed in KafkaProducer may be used elsewhere and should be closed by the KafkaController
+    }
+
+    private String sendForkliftWrappedMessage(String message, Map<String, String> messageProperties) throws ProducerException {
+        GenericRecord avroRecord = new GenericData.Record(forkliftSchema);
+        avroRecord.put(SCHEMA_FIELD_NAME_VALUE, message);
+        messageProperties = messageProperties == null ? new HashMap<>() : new HashMap<>(messageProperties);
+        //add the producer level properties but do not overwrite message level properties
+        for (Map.Entry<String, String> entry : this.properties.entrySet()) {
+            messageProperties.putIfAbsent(entry.getKey(), entry.getValue());
+        }
+        avroRecord.put(SCHEMA_FIELD_NAME_PROPERTIES, this.formatMap(messageProperties));
+        ProducerRecord record = new ProducerRecord<>(topic, null, avroRecord);
+        try {
+            RecordMetadata result = (RecordMetadata)kafkaProducer.send(record).get();
+            return result.topic() + "-" + result.partition() + "-" + result.offset();
+
+        } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+            throw new ProducerException("Error sending Kafka Message", e);
+        } catch (ExecutionException e) {
+            throw new ProducerException("Error sending Kafka Message", e);
+        }
+    }
+
+    private Schema addForkliftPropertiesToSchema(Schema schema) throws IOException {
+        String originalJson = schema.toString(false);
+        JsonNode propertiesField = mapper.readTree("{\"name\":\"forkliftProperties\",\"type\":\"string\",\"default\":\"\"," +
+                                                   "\"doc\":\"Properties added to support forklift interfaces. Format is key,value entries delimited with new lines\"}");
+        ObjectNode schemaNode = (ObjectNode)mapper.readTree(originalJson);
+        ArrayNode fieldsNode = (ArrayNode)schemaNode.get("fields");
+        fieldsNode.add(propertiesField);
+        schemaNode.set("fields", fieldsNode);
+        Schema.Parser parser = new Schema.Parser();
+        return parser.parse(mapper.writeValueAsString(schemaNode));
+    }
+
+    private GenericRecord addForkliftPropertiesToAvroObject(SpecificRecord message) throws IOException {
```

Comment:
A lot of what is going on here seems pretty inefficient. What's worse, this is done for every avro message we send out. Producing events is supposed to be fast.

I bet that the vast majority of our (future) use cases for a producer don't even use properties; could we partially optimize by not doing any of the hard work if we have no properties to insert?